warmrestart/warmRestartHelper.cpp

```diff
@@ -0,0 +1,472 @@
+
+    assert(getState() == WarmStart::RESTORED);
+
+    /*
+     * Iterate through all the entries in the recoveryMap and take note of the
+     * attributes associated to each.
+     */
+    auto it = m_recoveryMap.begin();
+    while (it != m_recoveryMap.end())
+    {
+        std::string key = it->first;
+        fvRecoveryMap fvMap = it->second;
+
+        int totalRecElems, staleRecElems, cleanRecElems, newRecElems, deleteRecElems;
+        totalRecElems = staleRecElems = cleanRecElems = newRecElems = deleteRecElems = 0;
+
+        std::vector<FieldValueTuple> fvVector;
+
+        for (auto itMap = fvMap.begin(); itMap != fvMap.end(); )
+        {
+            totalRecElems++;
+
+            auto recElem = itMap->first;
+            auto recElemState = itMap->second;
+
+            if (recElemState == STALE)
+            {
+                itMap = fvMap.erase(itMap);
+                staleRecElems++;
+            }
+            else if (recElemState == CLEAN)
+            {
+                cleanRecElems++;
+                transformKFV(recElem, fvVector);
+                ++itMap;
+            }
+            else if (recElemState == NEW)
+            {
+                newRecElems++;
+                transformKFV(recElem, fvVector);
+                ++itMap;
+            }
+            else if (recElemState == DELETE)
+            {
+                deleteRecElems++;
+                ++itMap;
+            }
+        }
+
+        if (staleRecElems == totalRecElems)
+        {
+            m_syncTable->del(key);
+            SWSS_LOG_NOTICE("Route reconciliation: deleting stale prefix %s\n",
+                            key.c_str());
+        }
+        else if (cleanRecElems == totalRecElems)
+        {
+            SWSS_LOG_NOTICE("Route reconciliation: no changes needed for existing"
+                            " prefix %s\n", key.c_str());
+        }
+        else if (newRecElems == totalRecElems)
+        {
+            m_syncTable->set(key, fvVector);
+            SWSS_LOG_NOTICE("Route reconciliation: creating new prefix %s\n",
+                            key.c_str());
+        }
+        else if (cleanRecElems)
+        {
+            m_syncTable->set(key, fvVector);
+            SWSS_LOG_NOTICE("Route reconciliation: updating attributes for prefix"
+                            " %s\n", key.c_str());
+        }
+        else if (newRecElems)
+        {
+            m_syncTable->set(key, fvVector);
+            SWSS_LOG_NOTICE("Route reconciliation: creating new attributes for "
+                            "prefix %s\n", key.c_str());
+        }
+        else if (deleteRecElems)
+        {
+            m_syncTable->del(key);
+            SWSS_LOG_NOTICE("Route reconciliation: deleting withdrawn prefix %s\n",
+                            key.c_str());
+        }
+
+        it = m_recoveryMap.erase(it);
+    }
+
+    /* Recovery map should be entirely empty by now */
+    assert(m_recoveryMap.size() == 0);
```

Comment:
We should recover gracefully here by falling back to cold restart.

Reply:
Not sure what do you mean by 'gracefully'. I understand that we should do something about it, but that's something the process-manager (supervisord) should take care of upon the app being cold-booted after the failed assertion.

Reply:
How will supervisord achieve a cold boot since warm restart is enabled? In any case, asserting is shouldn't be here in my opinion - that's way too harsh. Flushing the db and re-establishing the connection will cause the routes to re-download. That's why I mean by gracefully.